README.md
# Leaflet OverPass Layer

This plugin is a fork of https://github.com/GuillaumeAmat/leaflet-overpass-layer forked by https://github.com/kartenkarsten/leaflet-layer-overpass

## What is it?
A [Leaflet](http://leafletjs.com/) plugin to create a custom POI overlay - thanks to the [OSM](http://www.openstreetmap.org/) dataset and the [Overpass API](http://overpass-api.de/)

This fork will provide a more extensible version with more options and a deeper implementation of
overpass features.


## Installation

### NPM

```bash
$ npm install leaflet-overpass-layer
```

### Bower

If you use bower to install leaflet-layer-overpass, you need to load [JsClipper](https://github.com/mathisonian/JsClipper) first.

```bash
$ bower install leaflet-overpass-layer
```

## Usage

You can include and use the `OverpassLayer.css` and `OverpassLayer.bundle.js` files from the `dist` folder in your html.


```javascript
var attr_osm = 'Map data &copy; <a href="http://openstreetmap.org/">OpenStreetMap</a> contributors';
var attr_overpass = 'POI via <a href="http://www.overpass-api.de/">Overpass API</a>';

var osm = new L.TileLayer(
    'http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
    {
        'opacity': 0.7,
        'attribution': [attr_osm, attr_overpass].join(', ')
    }
);

var map = new L.Map('map')
.addLayer(osm)
.setView(new L.LatLng(52.265, 10.524), 14);


var opl = new L.OverPassLayer({

    'query': '(node({{bbox}})[organic];node({{bbox}})[second_hand];);out qt;',
});

map.addLayer(opl);
```
In order to get a valid query the [Overpass-turbo IDE](http://overpass-turbo.eu/) might help.

## What are the options?
You can specify an options object as an argument of L.OverPassLayer.
```javascript
options: {

  debug: false,
  minZoom: 15,
  endPoint: 'https://overpass-api.de/api/',
  query: '(node({{bbox}})[organic];node({{bbox}})[second_hand];);out qt;',
  loadedBounds: [],
  markerIcon: L.Icon(),
  timeout: 30 * 1000, // Milliseconds
  retryOnTimeout: false,
  noInitialRequest: false,
  minZoomIndicatorEnabled: true,
  minZoomIndicatorOptions: {
    position: 'topright',
    minZoomMessageNoLayer: 'No layer assigned',
    minZoomMessage: 'Current zoom level: CURRENTZOOM - All data at level: MINZOOMLEVEL'
},
  beforeRequest: function() {},
  afterRequest: function() {},
  onSuccess: function(data) {},
  onError: function(xhr) {},
  onTimeout: function(xhr) {},
};
```

## Dependencies

* [Leaflet](https://github.com/Leaflet/Leaflet)

## Development

*Warning: This fork use Git Flow to manage branches.*

In order to contribute to the project you should first clone the repository. The javascript source files
reside in the `src` folder and are concatenated and minified by Webpack in the `dist` folder. If you want to make changes
make them in the `src` folder and then run `npm run build` (or `npm run watch` for continuous build).
For that you first need to install all the needed packages for this project:
```
$ npm install
```


CHANGELOG.md
# Changelog

## 1.0.1
 * Correct some Type Definitions

## 1.0.0

* Initial version with added TypescriptDefinitions and support for drawing ways.

src/OverPassLayer.d.ts
// Type definitions for L-Overpass-Layer by GuillaumeAmat
// Project: https://github.com/GuillaumeAmat/leaflet-overpass-layer
// Definitions by: Oliver Schörwerth <https://github.com/grafzahl-io>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
// TypeScript Version: 2.3

import * as L from 'leaflet';

declare module 'leaflet' {
	interface OverPassLayer extends Layer {
	  options: OverPassLayerOptions;
	  initialize(options: OverPassLayerOptions): void;
	  _getPoiPopupHTML(tags: string[], id:string|number): HTMLElement;
	  _buildRequestBox(bounds: L.LatLngBounds): L.Rectangle;
	 	_addRequestBox(box: L.Layer): L.Layer;
	 	_getRequestBoxes(): void;
	 	_removeRequestBoxes(): L.LayerGroup;
	 	_addResponseBox(box: L.Layer): L.LayerGroup;
	 	_addResponseBoxes(requestBoxes: L.Layer[]): void;
	 	_isFullyLoadedBounds(bounds: L.LatLngBounds, loadedBounds: L.LatLngBounds): boolean;
	 	_getLoadedBounds(): L.LatLngBounds;
	 	_addLoadedBounds(bounds: L.LatLngBounds): void;
	 	/**
	 	 * @todo set return type
	 	 */
	 	_buildClipsFromBounds(bounds: L.LatLngBounds): any;
	 	 	/**
	 	 * @todo set types
	 	 */
	 	_buildBoundsFromClips(clis: any): any;
	 	_buildOverpassQueryFromQueryAndBounds(query: string,bounds: L.LatLngBounds): string;
	 	_buildOverpassUrlFromEndPointAndQuery(endPoint: string, query: string): string;
	 	_buildLargerBounds(bounds: L.LatLngBounds): L.LatLngBounds;
	 	 	/**
	 	 * @todo set types
	 	 */
	 	_setRequestInProgress(isInProgress: any): void;
	 	_isRequestInProgress(): any;
	 	_hasNextRequest(): boolean;
	 	 	/**
	 	 * @todo set types
	 	 */
	 	_getNextRequest(): any;
	 	_setNextRequest(nextRequest: any): void;
		_removeNextRequest(): void;
		 	 	/**
	 	 * @todo set types
	 	 */
		_prepareRequest(): any;
		 	 	/**
	 	 * @todo set types
	 	 */
		_sendRequest(bounds: L.LatLngBounds): any;
		_onRequestLoad(xhr: any, bounds: L.LatLngBounds): void;
		_onRequestTimeout(xhr: any, url: string, bounds: L.LatLngBounds): void;
		_onRequestLoadCallback(bounds: L.LatLngBounds): void;
		_onRequestCompleteCallback(): void;
    	onAdd(map: L.Map): this;
		onRemove(map: L.Map): this;
		setQuery(query: string): void;
		_resetData(): void;
		getData(): any;
	}

	class OverPassLayer {
		constructor(options?: OverPassLayerOptions);
	}

	interface OverPassLayerOptions {
		debug?: boolean;
		minZoom?: number;
		endpoint?: string;
		query: string;
		loadedBounds?: L.LatLngBounds; // could be also LatLngBoundsLiteral
		markerIcon?: L.Icon;
		timeout?: number;
		retryOnTimeout?: boolean;
		noInitialRequest?: boolean;
	  beforeRequest?(): void;
	  afterRequest?(): void;
	  /**
	   * @todo define data
	   */
	  onSuccess?(data: any): void;
	  onError?(): void;
	  onTimeout?(): void;
	  minZoomIndicatorEnabled?: boolean;
	  minZoomIndicatorOptions?: { minZoomMessageNoLayer: string, minZoomMessage: string };
	}
}

src/OverPassLayer.js
import L from 'leaflet';
import ClipperLib from 'js-clipper';
import './OverPassLayer.css';
import './MinZoomIndicator';

const OverPassLayer = L.FeatureGroup.extend({
  options: {
    debug: false,
    minZoom: 13,
    endPoint: 'https://overpass-api.de/api/',
    query: '(node({{bbox}})[organic];node({{bbox}})[second_hand];);out qt;',
    loadedBounds: [],
    markerIcon: null,
    timeout: 30 * 1000, // Milliseconds
    retryOnTimeout: false,
    noInitialRequest: false,

    beforeRequest() {},

    afterRequest() {},

    onSuccess(data) {
      for (let i = 0; i < data.elements.length; i++) {
        let pos;
        let wpos;
        let marker;
        const e = data.elements[i];

        if (e.id in this._ids) {
          continue;
        }

        this._ids[e.id] = true;

        if (e.type === 'node' && this.options.query.indexOf('node') !== -1) {
          pos = L.latLng(e.lat, e.lon);

          marker = this._getMarkerObject(pos, e);
          const popupContent = this._getPoiPopupHTML(e.tags, e.id);
          const popup = L.popup().setContent(popupContent);
          marker.bindPopup(popup);

          this._markers.addLayer(marker);
        } else if(e.type === 'way') {
          let waylist = [];
          if(e.geometry !== undefined) {
            for(let j = 0; j < e.geometry.length; j++) {
              wpos = L.latLng(e.geometry[j].lat, e.geometry[j].lon);
              this._ids[e.geometry[j].id] = true;
              waylist.push(wpos);
            }
            if(waylist.length > 0) {
             let firstpolyline = new L.Polyline(waylist, {
                color: '#53c0d7',
                weight: 3,
                opacity: 0.5,
                smoothFactor: 1

              });
              this._ways.addLayer(firstpolyline);
            }
          }
        } else {
          if(e.type !== 'node' && e.type !== 'way'
            && e.center !== undefined) {
            pos = L.latLng(e.center.lat, e.center.lon);
            /**
             * missing definition to 
             * set marker
             */
          }
        }
      }
    },

    onError() {},

    onTimeout() {},

    minZoomIndicatorEnabled: true,
    minZoomIndicatorOptions: {
      minZoomMessageNoLayer: 'No layer assigned',
      minZoomMessage:
        'Current zoom Level: CURRENTZOOM. Data are visible at Level: MINZOOMLEVEL.'
    }
  },

  initialize(options) {
    L.Util.setOptions(this, options);

    this._ids = {};
    this._loadedBounds = options.loadedBounds || [];
    this._requestInProgress = false;
  },

  _getMarkerObject(pos, e) {
    if (this.options.markerIcon) {
      return L.marker(pos, { icon: this.options.markerIcon });
    } else {
      return L.circle(pos, 20, {
        stroke: false,
        fillColor: '#af2560',
        fillOpacity: 1
      });
    }
  },

  _getPoiPopupHTML(tags, id) {
    let row;
    const link = document.createElement('a');
    const table = document.createElement('table');
    const div = document.createElement('div');

    link.href = `https://www.openstreetmap.org/edit?editor=id&node=${id}`;
    link.appendChild(document.createTextNode('Edit this entry in iD'));

    table.style.borderSpacing = '10px';
    table.style.borderCollapse = 'separate';

    for (const key in tags) {
      row = table.insertRow(0);
      row.insertCell(0).appendChild(document.createTextNode(key));
      row.insertCell(1).appendChild(document.createTextNode(tags[key]));
    }

    div.appendChild(link);
    div.appendChild(table);

    return div;
  },

  _buildRequestBox(bounds) {
    return L.rectangle(bounds, {
      bounds: bounds,
      color: '#204a87',
      stroke: false,
      fillOpacity: 0.1,
      clickable: false
    });
  },

  _addRequestBox(box) {
    return this._requestBoxes.addLayer(box);
  },

  _getRequestBoxes() {
    return this._requestBoxes.getLayers();
  },

  _removeRequestBox(box) {
    this._requestBoxes.removeLayer(box);
  },

  _removeRequestBoxes() {
    return this._requestBoxes.clearLayers();
  },

  _addResponseBox(box) {
    return this._responseBoxes.addLayer(box);
  },

  _addResponseBoxes(requestBoxes) {
    this._removeRequestBoxes();

    requestBoxes.forEach(box => {
      box.setStyle({
        color: 'black',
        weight: 2
      });
      this._addResponseBox(box);
    });
  },

  _isFullyLoadedBounds(bounds, loadedBounds) {
    if (loadedBounds.length === 0) {
      return false;
    }

    const subjectClips = this._buildClipsFromBounds([bounds]);
    const knownClips = this._buildClipsFromBounds(loadedBounds);
    const clipper = new ClipperLib.Clipper();
    const solutionPolyTree = new ClipperLib.PolyTree();

    clipper.AddPaths(subjectClips, ClipperLib.PolyType.ptSubject, true);
    clipper.AddPaths(knownClips, ClipperLib.PolyType.ptClip, true);

    clipper.Execute(
      ClipperLib.ClipType.ctDifference,
      solutionPolyTree,
      ClipperLib.PolyFillType.pftNonZero,
      ClipperLib.PolyFillType.pftNonZero
    );

    const solutionExPolygons = ClipperLib.JS.PolyTreeToExPolygons(
      solutionPolyTree
    );

    if (solutionExPolygons.length === 0) {
      return true;
    } else {
      return false;
    }
  },

  _getLoadedBounds() {
    return this._loadedBounds;
  },

  _addLoadedBounds(bounds) {
    this._loadedBounds.push(bounds);
  },

  _buildClipsFromBounds(bounds) {
    return bounds.map(bound => [
      {
        X: bound._southWest.lng * 1000000,
        Y: bound._southWest.lat * 1000000
      },
      {
        X: bound._southWest.lng * 1000000,
        Y: bound._northEast.lat * 1000000
      },
      {
        X: bound._northEast.lng * 1000000,
        Y: bound._northEast.lat * 1000000
      },
      {
        X: bound._northEast.lng * 1000000,
        Y: bound._southWest.lat * 1000000
      }
    ]);
  },

  _buildBoundsFromClips(clips) {
    return clips.map(clip =>
      L.latLngBounds(
        L.latLng(clip[0].Y / 1000000, clip[0].X / 1000000).wrap(),
        L.latLng(clip[2].Y / 1000000, clip[2].X / 1000000).wrap()
      )
    );
  },

  _buildOverpassQueryFromQueryAndBounds(query, bounds) {
    const sw = bounds._southWest;
    const ne = bounds._northEast;
    const coordinates = [sw.lat, sw.lng, ne.lat, ne.lng].join(',');

    query = query.replace(/(\/\/.*)/g, '');
    query = query.replace(/(\{\{bbox\}\})/g, coordinates);

    return query;
  },

  _buildOverpassUrlFromEndPointAndQuery(endPoint, query) {
    return `${endPoint}interpreter?data=[out:json];${query}`;
  },

  _buildLargerBounds(bounds) {
    const width = Math.abs(bounds._northEast.lng - bounds._southWest.lng);
    const height = Math.abs(bounds._northEast.lat - bounds._southWest.lat);
    const biggestDimension = width > height ? width : height;

    bounds._southWest.lat -= biggestDimension / 2;
    bounds._southWest.lng -= biggestDimension / 2;
    bounds._northEast.lat += biggestDimension / 2;
    bounds._northEast.lng += biggestDimension / 2;

    return L.latLngBounds(
      L.latLng(bounds._southWest.lat, bounds._southWest.lng).wrap(),
      L.latLng(bounds._northEast.lat, bounds._northEast.lng).wrap()
    );
  },

  _setRequestInProgress(isInProgress) {
    this._requestInProgress = isInProgress;
  },

  _isRequestInProgress() {
    return this._requestInProgress;
  },

  _hasNextRequest() {
    if (this._nextRequest) {
      return true;
    }

    return false;
  },

  _getNextRequest() {
    return this._nextRequest;
  },

  _setNextRequest(nextRequest) {
    this._nextRequest = nextRequest;
  },

  _removeNextRequest() {
    this._nextRequest = null;
  },

  _prepareRequest() {
    if (this._map.getZoom() < this.options.minZoom) {
      return false;
    }

    const bounds = this._buildLargerBounds(this._map.getBounds());
    const nextRequest = this._sendRequest.bind(this, bounds);

    if (this._isRequestInProgress()) {
      this._setNextRequest(nextRequest);
    } else {
      this._removeNextRequest();
      nextRequest();
    }
  },

  _sendRequest(bounds) {
    const loadedBounds = this._getLoadedBounds();

    if (this._isFullyLoadedBounds(bounds, loadedBounds)) {
      this._setRequestInProgress(false);
      return;
    }

    const requestBounds = this._buildLargerBounds(bounds);
    const url = this._buildOverpassUrlFromEndPointAndQuery(
      this.options.endPoint,
      this._buildOverpassQueryFromQueryAndBounds(
        this.options.query,
        requestBounds
      )
    );
    const request = new XMLHttpRequest();
    const beforeRequestResult = this.options.beforeRequest.call(this);

    if (beforeRequestResult === false) {
      this.options.afterRequest.call(this);

      return;
    }

    this._setRequestInProgress(true);

    if (this.options.debug) {
      this._addRequestBox(this._buildRequestBox(requestBounds));
    }

    request.open('GET', url, true);
    request.timeout = this.options.timeout;

    request.ontimeout = () =>
      this._onRequestTimeout(request, url, requestBounds);
    request.onload = () => this._onRequestLoad(request, requestBounds);

    request.send();
  },

  _onRequestLoad(xhr, bounds) {
    if (xhr.status >= 200 && xhr.status < 400) {
      this.options.onSuccess.call(this, JSON.parse(xhr.response));

      this._onRequestLoadCallback(bounds);
    } else {
      this._onRequestErrorCallback(bounds);

      this.options.onError.call(this, xhr);
    }

    this._onRequestCompleteCallback(bounds);
  },

  _onRequestTimeout(xhr, url, bounds) {
    this.options.onTimeout.call(this, xhr);

    if (this.options.retryOnTimeout) {
      this._sendRequest(url);
    } else {
      this._onRequestErrorCallback(bounds);
      this._onRequestCompleteCallback(bounds);
    }
  },

  _onRequestLoadCallback(bounds) {
    this._addLoadedBounds(bounds);

    if (this.options.debug) {
      this._addResponseBoxes(this._getRequestBoxes());
    }
  },

  _onRequestErrorCallback(bounds) {
    if (this.options.debug) {
      this._removeRequestBox(this._buildRequestBox(bounds));
    }
  },

  _onRequestCompleteCallback() {
    this.options.afterRequest.call(this);

    if (this._hasNextRequest()) {
      const nextRequest = this._getNextRequest();

      this._removeNextRequest();

      nextRequest();
    } else {
      this._setRequestInProgress(false);
    }
  },

  onAdd(map) {
    this._map = map;

    if (this.options.minZoomIndicatorEnabled === true) {
      if (this._map.zoomIndicator) {
        this._zoomControl = this._map.zoomIndicator;
        this._zoomControl._addLayer(this);
      } else {
        this._zoomControl = new L.Control.MinZoomIndicator(
          this.options.minZoomIndicatorOptions
        );

        this._map.addControl(this._zoomControl);

        this._zoomControl._addLayer(this);
      }
    }

    if (this.options.debug) {
      this._requestBoxes = L.featureGroup().addTo(this._map);
      this._responseBoxes = L.featureGroup().addTo(this._map);
    }

    this._markers = L.featureGroup().addTo(this._map);

    this._ways = L.featureGroup().addTo(this._map);

    if (!this.options.noInitialRequest) {
      this._prepareRequest();
    }

    this._map.on('moveend', this._prepareRequest, this);
  },

  onRemove(map) {
    L.LayerGroup.prototype.onRemove.call(this, map);

    this._resetData();

    map.off('moveend', this._prepareRequest, this);

    this._map = null;
  },

  setQuery(query) {
    this.options.query = query;
    this._resetData();
    this._prepareRequest();
  },

  _resetData() {
    this._ids = {};
    this._loadedBounds = [];
    this._requestInProgress = false;

    this._markers.clearLayers();

    if (this.options.debug) {
      this._requestBoxes.clearLayers();
      this._responseBoxes.clearLayers();
    }
  },

  getData() {
    return this._data;
  }
});

L.OverPassLayer = OverPassLayer;
L.overpassLayer = options => new L.OverPassLayer(options);

export default OverPassLayer;
